# Lab 0.1 — Course Environment and Data Tour

*Course 1258 — Applied AI for Government IT Professionals · Ch00 Course
Launch and Lab Environment · 20 minutes · CloudShare VM (JupyterLab) ·
OpenAI API via `lab_common` (canned fallback when no key)*

You have a working VM. Before the course leans on it, take a few minutes to
learn the room: where the notebooks are, where the data is, and how a
notebook reaches the model.

## Objectives

By the end of this lab, you will:

- Navigate JupyterLab confidently: cells, run order, kernel restart.
- Confirm the six shipped datasets load, and know what each one is for.
- Make one successful call to the OpenAI API using the key from `.env`.

## Setup

- **Datasets:** all six course datasets in `data/` (provenance in
  `data/MANIFEST.json`); this lab touches `federal_ai_use_cases.csv` and
  `chicago_311.csv` directly.
- **API key:** the environment cell below imports `lab_common`, whose loader
  reads `OPENAI_API_KEY` from the environment (classroom VM) or from the
  course `.env` file. It prints the `.env` path and model alias — **never
  the key**. With no key, the model cell uses a realistic canned reply, so
  the tour completes either way.
- **The one rule that never changes:** public or synthetic data only. Never
  paste real agency, personal or sensitive data into any assistant.
- Cells marked `# YOUR CODE` are for you to complete. Each has a reference
  fallback that runs if you leave it blank, so the notebook never gets stuck
  — replace the fallback with your own line when you can.

## Steps

1. **(1 min)** Open `lab_0.1_environment_tour.ipynb` — you are here. Run
   cells top to bottom (Shift+Enter), or **Run → Run All Cells** now and
   read along.
2. **(3 min)** Run the **environment** cell. Confirm it prints the `.env`
   path it loaded and the model alias — but never the key itself.
3. **(3 min)** Run the **data inventory** cell: it reads
   `data/MANIFEST.json` and prints each dataset with its row count and
   one-line description.
4. **(3 min)** Load `federal_ai_use_cases.csv` with pandas. Print `.shape`
   and `.columns`.
5. **(2 min)** Answer in the notebook: how many use cases are flagged
   `is_high_impact`?
6. **(3 min)** Load `chicago_311.csv` and print the five most common
   `sr_type` values.
7. **(3 min)** Run the **first API call** cell: a one-sentence prompt,
   printing the reply and the token counts.
8. **(2 min)** Deliberately break something: restart the kernel, re-run only
   the last cell, and read the error. Then **Run All** to recover.

### Step 2 — The environment cell (provided)

This cell finds the `labs/` folder, imports `lab_common` (which loads the
`.env` file), and prints where the key came from — **never the key itself**.
On the VM the key is injected for you; off the VM, `lab_common` reads it from
a `.env` file at the course root.

In [1]:
# Locate the labs/ folder no matter where Jupyter was started from.
import os, sys
from pathlib import Path

for _cand in (Path.cwd(), *Path.cwd().parents):
    if (_cand / "lab_common.py").is_file():
        os.chdir(_cand)
        if str(_cand) not in sys.path:
            sys.path.insert(0, str(_cand))
        break

import lab_common as lc

print(".env loaded from:", lc.DOTENV_PATH or "(none — relying on environment variables)")
print("chat model alias  :", lc.CHAT_MODEL)
print("embedding model   :", lc.EMBED_MODEL)
print("API key configured:", bool(lc.get_client()))
print("data directory    :", lc.DATA_DIR)

.env loaded from: (none — relying on environment variables)
chat model alias  : gpt-5-mini
embedding model   : text-embedding-3-small
API key configured: False
data directory    : /Users/iahmad/Creator/Courses_and_conferences/LT/1258/1258a4-author-input/labs/data


### Step 3 — Data inventory (provided)

`data/MANIFEST.json` says where each dataset came from and what it is for.
This cell prints each file with its row count and one-line description.

In [2]:
import json
import pandas as pd

manifest = json.loads((lc.DATA_DIR / "MANIFEST.json").read_text())
print(f"data pack retrieved: {manifest['retrieved']}\n")
for ds in manifest["datasets"]:
    n = len(pd.read_csv(lc.DATA_DIR / ds["file"], low_memory=False))
    print(f"{ds['file']:28s} {n:>6,} rows — {ds['description']}")

data pack retrieved: 2026-08-01

federal_ai_use_cases.csv      1,200 rows — OMB 2025 Federal Agency AI Use Case Inventory — individually reported use cases (3,611 across 56 agencies; capped here). The course's own subject matter, as data.
federal_ai_cots.csv             900 rows — OMB 2025 inventory — consolidated commercial off-the-shelf AI use cases.
chicago_311.csv               4,000 rows — City of Chicago 311 service requests (Socrata). Deliberately messy real data: blank fields, inconsistent casing, duplicate rows — the data-quality lab.
cdc_flu_wastewater.csv        3,000 rows — CDC wastewater surveillance for Influenza A — time series by site. The real version of the early-warning story told in Chapter 2.
epa_aqi_by_county.csv           997 rows — EPA Annual Air Quality Index by county, 2024 — clean, small, national. Good for grouping, ranking and charting.
nyc_air_quality.csv           3,000 rows — NYC air quality surveillance (PM2.5 and related indicators) by neighbourhood an

### Step 4 — Load the federal AI use case inventory

This CSV is the course's own subject matter as data: every AI use case the
agencies reported to OMB. Load it and print its shape and columns.

In [3]:
uc = None
# YOUR CODE: read data/federal_ai_use_cases.csv into `uc` with pd.read_csv
# (lc.DATA_DIR points at the data/ folder; use encoding="utf-8-sig" — the file
# starts with a byte-order mark). Then print uc.shape and list(uc.columns).

if uc is None:
    uc = pd.read_csv(lc.DATA_DIR / "federal_ai_use_cases.csv", encoding="utf-8-sig")
    print("(reference load applied — replace with your own line above)\n")
print("shape:", uc.shape)
print("columns:", list(uc.columns))

(reference load applied — replace with your own line above)

shape: (1200, 36)
columns: ['agency', 'agency_name', 'id', 'use_case_name', 'agency_bureau', 'contact_email', 'is_withheld', 'development_stage', 'is_high_impact', 'HI_justification', 'topic_area', 'classification', 'problem_solved', 'benefits', 'system_outputs', 'operational_date', 'contracting_usage', 'vendor_name', 'have_ato', 'system_name_ato', 'data_description', 'link_to_data', 'has_pii', 'pia_url', 'demographic_features', 'has_custom_code', 'code_url', 'hi_testing_conducted', 'hi_assessment_completed', 'hi_potential_impacts', 'hi_independent_review', 'hi_ongoing_monitoring', 'hi_training_established', 'hi_failsafe_presence', 'hi_appeal_process', 'hi_public_consultation']


### Step 5 — How many use cases are flagged high-impact?

The `is_high_impact` column answers OMB's risk question. Count how many use
cases are flagged `High-impact` — and notice what *other* values the column
contains while you are at it.

In [4]:
high_impact_count = None
# YOUR CODE: count rows where uc["is_high_impact"] == "High-impact".

if high_impact_count is None:
    high_impact_count = int((uc["is_high_impact"] == "High-impact").sum())
    print("(reference count applied)\n")
print(uc["is_high_impact"].value_counts().to_string())
print(f"\nflagged High-impact: {high_impact_count} of {len(uc)}")

(reference count applied)

is_high_impact
Not High-impact                              752
High-impact                                  143
Presumed High-Impact, but Not High-impact     79

flagged High-impact: 143 of 1200


### Step 6 — Chicago 311: the five most common request types

`chicago_311.csv` is a recent extract of the City's 311 service requests —
deliberately real and messy. You will meet it again in Labs 5.2, 6.1 and 6.2.

In [5]:
top5 = None
# YOUR CODE: load data/chicago_311.csv (low_memory=False) and store the five
# most common values of the "sr_type" column in `top5`.

if top5 is None:
    _c311 = pd.read_csv(lc.DATA_DIR / "chicago_311.csv", low_memory=False)
    top5 = _c311["sr_type"].value_counts().head(5)
    print("(reference load applied)\n")
print(top5.to_string())

(reference load applied)

sr_type
Aircraft Noise Complaint     1547
311 INFORMATION ONLY CALL     760
Water On Street Complaint     257
Tree Emergency                159
Graffiti Removal Request      136


### Step 7 — Your first API call

One sentence in, one sentence out, and read the token counts — the unit you
pay in. If no key is configured you get a canned reply so the tour still
works; the token line tells you it is offline.

In [6]:
CANNED_REPLY = (
    "The OMB AI use case inventory tracks how federal agencies are using "
    "artificial intelligence — each reported system's purpose, development "
    "stage, and whether it is high-impact."
)

client = lc.get_client()
if client is not None:
    resp = client.chat.completions.create(
        model=lc.CHAT_MODEL,
        messages=[{"role": "user", "content":
                   "In one sentence, what does the OMB federal AI use case "
                   "inventory track?"}],
    )
    print(resp.choices[0].message.content)
    print(f"\ntokens — prompt: {resp.usage.prompt_tokens}, "
          f"completion: {resp.usage.completion_tokens}, "
          f"total: {resp.usage.total_tokens}")
else:
    print(lc.chat([{"role": "user", "content": "..."}], offline=CANNED_REPLY))
    print("\n(offline canned reply — token counts need a live call; "
          "a call this size is typically ~40 tokens total)")

The OMB AI use case inventory tracks how federal agencies are using artificial intelligence — each reported system's purpose, development stage, and whether it is high-impact.

(offline canned reply — token counts need a live call; a call this size is typically ~40 tokens total)


### Step 8 — Break it on purpose

1. **Kernel menu → Restart Kernel** (the circular arrow). This wipes every
   variable.
2. Re-run *only* this cell. Read the `NameError`: `lc` no longer exists
   because the import cell above has not run in the new kernel.
3. **Run menu → Run All Cells** to recover.

This is the single most common notebook confusion in class — make it happen
now, on purpose, when nothing is at stake.

In [7]:
print("lc is still defined:", "lc" in dir(), "| uc rows:", len(uc))

lc is still defined: True | uc rows: 1200


## Deliverable

A fully run notebook (Run All, top to bottom, no errors) with:
1. The data inventory printed in Step 3.
2. Your high-impact count from Step 5.
3. The top-5 request types from Step 6.
4. The model's reply and token counts from Step 7 (or the offline note).

## Reflection

1. Why did the last cell fail after a kernel restart?
2. Which of the six datasets would you reach for to answer "which agencies
   use AI for document processing"?
3. What did the token counts suggest about cost at agency scale?

## Debrief (instructor-led)

- Show of hands: who hit the `NameError` in Step 8 before Run All fixed it?
  Say it aloud: kernel state lives in memory, not in the file — "Restart &
  Run All" is the notebook's clean build.
- Look at the `is_high_impact` value counts together: what on earth is
  `Presumed High-Impact, but Not High-impact`? (Matched OMB's checklist on
  paper; the agency judged it not actually high-impact. Real government data
  has labels like this.)
- One minute on the token counts: ~40 tokens for a sentence. What does a
  50,000-document month do to that number? (Preview of Lab 7.1's cost
  exercise.)

## Troubleshooting

- **`NameError: name 'lc' is not defined`.** You restarted the kernel or ran
  a cell out of order — **Run → Run All Cells**.
- **`FileNotFoundError: data/...`.** You are not running from the `labs/`
  folder; the Step 2 setup cell should have fixed this — re-run from the top.
- **`UnicodeDecodeError` or a column named `\ufeff...`.** The federal CSV has
  a byte-order mark; load it with `encoding="utf-8-sig"`.
- **Step 7 prints the canned reply.** No key is configured — on the class VM,
  tell your instructor; off the VM this is expected and the lab still works.
- **The data-inventory cell is slow.** It reads all six CSVs end to end to
  count rows — 20–30 seconds is normal on the VM.